### PR TITLE
[RFR] Get the scoreboard of a grid

### PR DESCRIPTION
**Pull request by benjaminpayoux**
[Trello](https://trello.com/c/pxp0PSNR/1-1-as-peter-given-a-small-board-with-a-single-destroyer-i-want-to-know-every-cell-probability-to-have-the-boat-on-it)

## TODO
- [X] Tests different ships size on 3x3 grid
- [X] Display the scoreboards along tests

## Result
last

### CODE REVIEW

File: test/battleship/game/scoreboard_test.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+	// When computing its possible positions
+	// on a 3x3 grid
+	scoreBoard := grid.GetScoreBoard(3, ship)
+
+	// Then it should equals this score board
+	expected := "2 3 2\n3 4 3\n2 3 2"
+
+	actual := scoreboard.GetStringFromScoreBoard(scoreBoard)
+
+	if actual != expected {
```
Comment (JulienMattiussi) [inline test/battleship/game/scoreboard_test.go]: You should use something like this to be more lisible
https://godoc.org/github.com/stretchr/testify/assert
https://github.com/corbym/gocrest

File: test/battleship/game/scoreboard_test.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+	"battleship/scoreboard"
+	"fmt"
+	"testing"
+)
+
+func TestScoreBoardWithOneShip(t *testing.T) {
+	// Given one 2 long ship
+	ship := game.Ship{2}
+
+	// When computing its possible positions
```
Comment (JulienMattiussi) [inline test/battleship/game/scoreboard_test.go:48]: This is also part of "Given" step

File: test/battleship/game/scoreboard_test.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+import (
+	"battleship/game"
+	"battleship/grid"
+	"battleship/scoreboard"
+	"fmt"
+	"testing"
+)
+
+func TestScoreBoardWithOneShip(t *testing.T) {
```
Comment (JulienMattiussi) [inline test/battleship/game/scoreboard_test.go]: This should identify the tested function like
"TestGetScoreBoard"

File: src/battleship/grid/grid.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+	if ship.Length > gridSize {
+		return nil, errors.New("The ship is too long. Its length must be inferior to the grid size.")
+	}
+
+	scoreBoard := scoreboard.MakeScoreBoard(gridSize)
+
+	for row := 0; row < gridSize; row++ {
+		for column := 0; column < gridSize; column++ {
+
+			if column+ship.Length <= gridSize {
```
Comment (mchaffotte) [inline src/battleship/grid/grid.go:38]: Did you run the command to lint the code ? there is no space around the +

File: src/battleship/scoreboard/scoreboard.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+	squares := make([][]int, size)
+
+	for i := 0; i < size; i++ {
+		squares[i] = make([]int, size)
+	}
+
+	return ScoreBoard{size, squares}
+}
+
+func GetStringFromScoreBoard(scoreBoard *ScoreBoard) string {
```
Comment (mchaffotte) [inline src/battleship/scoreboard/scoreboard.go]: maybe toString is enough as function name as we know we are using an scoreboard

File: src/battleship/scoreboard/scoreboard.go [hunk 6]
```diff
@@ -0,0 +1,36 @@
+package scoreboard
+
+import "strconv"
+
+type ScoreBoard struct {
+	Length  int
+	Squares [][]int
+}
+
+func MakeScoreBoard(size int) ScoreBoard {
```
Comment (mchaffotte) [inline src/battleship/scoreboard/scoreboard.go]: you use NewGrid in previous file, why did you change the name in that file MakeScoreBoard ? Be consistent

File: test/battleship/game/scoreboard_test.go [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+	expected := &scoreboard.ScoreBoard{3, squares}
+
+	// When computing its possible positions
+	// on a 3x3 grid
+	actual, _ := grid.GetScoreBoard(3, ship) // adresse récupérée
+
+	// Then it should equals this score board
+	then.AssertThat(t, actual, is.EqualTo(expected).Reason("1 square long ship on 3x3 grid"))
+	DisplayScoreBoard(actual, ship)
```
Comment (mchaffotte) [inline test/battleship/game/scoreboard_test.go]: You can reduce the visibility of that method as it is used only in that file

**Review by mchaffotte — CHANGES_REQUESTED** [review]

File: test/battleship/game/scoreboard_test.go [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+func TestGetScoreBoardShouldNotBeComputableWithTooLongShip(t *testing.T) {
+	// Given one 4 squares long ship
+	ship := game.Ship{4}
+
+	// When computing its possible positions
+	// on a 3x3 grid
+	_, error := grid.GetScoreBoard(3, ship)
+
+	// Then it should not be computed and return an error
+	if error == nil {
```
Comment (JulienMattiussi) [inline test/battleship/game/scoreboard_test.go]: Why doesn't this case just return a zero filled scoreBoard ?

File: src/battleship/scoreboard/scoreboard.go (hunk 6 above)
Comment (JulienMattiussi) [reply src/battleship/scoreboard/scoreboard.go]: For me it ok, Grid was the playing board
The score Board is different
Is'nt this ?

File: src/battleship/grid/grid.go (hunk 4 above)
Comment (benjaminpayoux) [reply src/battleship/grid/grid.go:38]: The vscode extension auto format this line and the `make lint` does not complain about it.

File: test/battleship/game/scoreboard_test.go (hunk 8 above)
Comment (benjaminpayoux) [reply test/battleship/game/scoreboard_test.go]: Right. A too long ship will not be computed but a score board can still be returned

File: src/battleship/scoreboard/scoreboard.go (hunk 6 above)
Comment (benjaminpayoux) [reply src/battleship/scoreboard/scoreboard.go]: Even if the models are different, the methods do the same thing, which is creating a two dimensionnal array. So I agree to stay consistent on methods name.

**Review by mchaffotte — APPROVED** [review]
